WordPress/gutenberg
# Extensibility: Define customClassname behavior as filtered blocks support

This follows the work done in #3318 and uses the same extensibility APIs to define the custom className behavior.

Unlike the anchor, this customClassname is opt-out. to opt-out of if, you need to add the following code to the block settings:

```js
supports: {
  customClassName: false
}
```

This surfaces some issues with the anchor definition:

 - This update the `extraProps` filter by passing the original element props to the first filter. This is necessary because we have to append the custom className to the existing className prop.

 - we can't call  `cloneElement` on element fragments which break the anchor/classname filters if you use both on a single block. To resolve this I introduced a `extendElement` helper. It can be improved IMO, open to suggestions on how to handle this better.

 - This also separates the `generatedClassName` from the `customClassName` to allow simple blocks like heading, paragraphs to support custom class names without having a generated one which was not possible. The `generatedClassName` bahavior could also be refactored to use the extensibility API.

 - This removes the "advanced controls" panel, We could provide it back by grouping these two extensions into a unique one. 

**Testing instructions**

 - Try adding a custom classname on a heading block
 - Try an anchor on the heading block.
 - Save and refresh.
 - You should still see these values populated properly
